Yogev/enhance debugging

yogev-lb opened the pull request:
* Added some minor changes to the package in order to be able to invoke additional API calls that were not supported.
* Added a String() method that makes it easier to debug the IF_SEND/IF_RECV calls (not clean but eases the debug)
* Added IsGlobal API for exposing the isGlobal Range property.
* Fix polling mechanism since some of the devices operations timeout on the default tmo set.

adir-lb commented on pkg/core/method.go:
```diff
@@ -221,18 +221,29 @@ func (m *MethodCall) Execute(c CommunicationIntf, proto drive.SecurityProtocol,
+			if len(resp) > 0 {
+				cont = false
+			}
+		case <-timeoutC:
```
@yogev-lb just to understand -
the old code did have timeout (100 iterations X 10ms interval = 1s timeout). what you are saying is he default session timeout wasn't taken into account? 
can you explain the purpose of this timeout (better add such explanation to the commit message)?

yogev-lb replied:
@adir-lb that is exectily what im saying
some operations - like revert for intel drives took around 5s
this caused a timeout.
the spec talks about a default timeout of 30s i think so i set it to this tmo value by default.

adir-lb commented on pkg/core/session.go:
```diff
@@ -27,6 +27,7 @@ var (
 	ErrPropertiesCallFailed        = errors.New("the properties call returned non-zero")
```
@yogev-lb can you elaborate in commit message why you had to move the definitions?

adir-lb commented on pkg/core/method.go:
```diff
@@ -209,6 +419,7 @@ func (m *MethodCall) Execute(c CommunicationIntf, proto drive.SecurityProtocol,
 		return nil, ErrReceivedUnexpectedResponse
 	}
 
+	// log.Printf("exec cmd: %s", m.String(b))
```
let's remove this comment.

adir-lb commented on pkg/core/method.go:
```diff
@@ -85,9 +204,100 @@ func NewMethodCall(iid InvokingID, mid MethodID, flags MethodFlag) *MethodCall {
+		}
+		_, ok4 := param.([]byte)
+		if ok4 {
+			//log.Printf("got []byte on params[%d]", i)
```
don't forget to remove the code in comment.

adir-lb commented on pkg/core/method.go:
```diff
@@ -85,9 +204,100 @@ func NewMethodCall(iid InvokingID, mid MethodID, flags MethodFlag) *MethodCall {
+	}
+}
+
+func (m *MethodCall) String(buf []byte) string {
```
why this method gets `buf` param?

yogev-lb replied:
removed

yogev-lb commented:
after looking today at upstream repo i see that there are many tasks that are started to be implemeneted that we implemented in our library:

for example they do a big refactor in this PR:
* https://github.com/open-source-firmware/go-tcg-storage/pull/88

Also they add get-msid and set-sid to the library which i implemented in the wrapping sed library here:
* https://github.com/open-source-firmware/go-tcg-storage/issues/89

I may wait with the upstream PR after the https://github.com/open-source-firmware/go-tcg-storage/pull/88 is merged
to see what are the leftovers to merge.

ChriMarMe commented:
Hello there, I wrapped up the refactor PR in upstream repo. TakeOwnership will take more time and rework of sessions as well. 

A review would be helpful to get the changes in asap, so you can continue with your PR. (I also would be happy to see a PR of these changes in the upstream repo ;)

ChriMarMe commented:
Hello, me again.

https://github.com/open-source-firmware/go-tcg-storage/pull/94 

This implements take ownership and SetMBREnable. 

It would be awesome if you can provide me with some feedback. :)

ChriMarMe commented on pkg/core/session.go:
```diff
@@ -237,6 +237,7 @@ func NewControlSession(d drive.DriveIntf, d0 *Level0Discovery, opts ...ControlSe
 	// Technically we should be able to advertise 0 here and the disk should pick
 	// for us, but that results in small values being picked in practice.
 	rhp.MaxComPacketSize = 1024 * 1024 // 1 MiB for good measure
+	//rhp.MaxComPacketSize = 1024 * 32 // LBM1-20784
```
Is that required? Until now the only usecase for MaxComPacketSize is when transmitting PBA image to MBR, but that doesnt effect recieving capabilities, which is handled by HostProperties ( at least that is the way I understand it)

Also check out "takeOwnership" branch in upstream repo. 

Maybe it is benefitial for all of us if we can work on upstream together.

lev-lb replied:
@ChriMarMe : please see the [commit message](https://github.com/LightBitsLabs/go-tcg-storage/pull/1/commits/5e90e664891416fcb463cd0ea42566bcbd4f763f) that introduces this change. we're getting sporadic errors on at least one SED model when using such huge buffers. so far, dropping it to, IIRC, pretty much anything below 1MB seemed to fix the issue for that particular model, but we're not taking any chances.

it might be argued that the problem we're seeing is peculiar to a certain server model (PCIe topology with switches) and a certain SED model, and so this might be addressed with a model-specific "quirk", but i personally am not very fond of those. we might also be able to do a bit better than hard-coding a value (any value, really, 32kB is no more or less magical than 1MB) while keeping it generic. this is our current brute-force hack. "fix it twice", and all that... ;)

ChriMarMe replied:
@lev-lb  Hmmm.....yeah, we can introduce checking of properties and let some code either try and error until working values are found or set boundary values ourselves.

Ok, this is not a urgend issue for me, lets wait what the device vendor says. 

Let me know :) I'm really curious

ChriMarMe replied:
I would not consider the process of exchanging properties as negotiation, but maybe there is something in the spec I dont get. What do I miss?

ChriMarMe replied:
Ok, that is bonkers xD

ChriMarMe replied:
```
5.2.2.3 Setting HostProperties
If the method is invoked with the optional HostProperties parameter, the list of name/value pairs that the
TPer MAY support is the list of properties inTable 168.

These values MAY be submitted in any order by the host. Not all values are required to be submitted.
Subsequent submission of these values (in a subsequent invocation of the Properties method)
SHALL supersede values submitted to previous invocations of the Properties method for that ComID.

Submitted values, if applicable, SHALL only apply to sessions started after the submission of those
values, and not to sessions that are already open on that ComID.

The TPer uses these host properties when it is constructing responses to be transmitted to the host.
The host MAY omit properties as necessary, depending on the host’s communications capabilities.

If the host omits a property, the value of that property SHALL NOT change from its current value. If the
host specifies a value for a property that does not meet the minimum requirement as defined in Table
168, then the TPer SHALL use the minimum value defined in Table 168 in place of the value supplied
by the host.
These values reflect the cumulative modifications of all processed Properties methods for the
associated ComID.

If the host sends the HostProperties parameter in its Properties invocation, the TPer SHALL respond
with ALL host properties it supports, with their current values, in the HostProperties parameter of its
method response.
If a host includes property parameters to the Properties method invocation that the TPer does not
support, the TPer SHALL ignore those parameters, and SHALL NOT return them in its response.
It is the host's responsibility to insure that Properties method invocations have processed prior to
invocation of any session startup methods that rely on those invocations. Values for HostProperties at
session startup rely on the Properties method invocations that have been processed by the TPer.
```

This looks to me like the exact opposite. That's why I'm confused.

lev-lb replied:
@ChriMarMe : sorry, not sure i'm following. what looks like the opposite of what? both passages quoted appear to be reasonably consistent with one another to me. care to point out the specific contradiction?

ChriMarMe replied:
Opposite was not quite the right wording on my side. 

I read it the way that the TPer uses the Values supplied in HostProperties parameter. So responsibility to get the values correct is on the hosts side.

What I understood from your former quote was:
TPer has some fixed values for HostProperties which needs to be reported to the host.

lev-lb replied:
@ChriMarMe : 

> I read it the way that the TPer uses the Values supplied in HostProperties parameter. So responsibility to get the values correct is on the hosts side.

i don't think this is a viable interpretation. TPers (SEDs) have limited HW resources and plenty of features are optional, so i don't think the host can dictate to TPer session params outside the scope of TPer's capabilities. besides, if this was so - why even bother with reporting back the TPer properties separately?

> What I understood from your former quote was: TPer has some fixed values for HostProperties which needs to be reported to the host.

i don't think that's it either. my interpretation is that the TPer has a set of capabilities (inc. max buffer sizes) that _it_ supports. it reports those in the _response_ `SMUID.Properties.Properties` (see 5.2.2.1.2.1). they are "fixed" insofar as they're limited by the SED's HW & FW, and they don't change depending on the host that's sitting on the other side of the wire or the values that host might have specified, but they are _not_ `HostProperties`.

the _response_ `SMUID.Properties.HostProperties` (5.2.2.1.2.2), on the other hand, is what the TPer proclaims it'll use for a given set of sessions (q.v ComID). these are _not_ fixed, and host SW passing in different `HostProperties` in the request might have different corresponding values returned, at TPer's discretion (within the bounds of the capabilities of both sides, i would imagine). this corresponds to the following two passages that me and you quoted, respectively:

> this portion of the method result SHALL include the communications limitations and capabilities that **the TPer SHALL use** for messages sent from the TPer to the Host.

> the TPer SHALL respond with ALL host properties it supports, with **their current values**, in the HostProperties parameter of its method response

in the meantime we were able to reproduce the issue we're seeing with the SED model in question with plenty of other SEDs of the same model, on several different server chassis of different models by different vendors with a few different Linux kernel versions. so it seems to be a vendor specific issue or a SED model specific one.

ChriMarMe replied:
Ok, now I got it. Thank you for taking the time to explain it to me. 

Great work! :+1:

lev-lb replied:
do keep in mind that i'm not one of the spec co-authors, so it's just my interpretation. empirically, it does appear to match the interpretation of most of the NVMe SED vendors we were able to lay our hands on so far - but not this one particular vendor, it seems. :)

i'll keep this thread posted once we hear back from their team.

lev-lb replied:
@ChriMarMe : after the 1st phase of root-causing the issue the vendor's engineering team seems to agree with my take on the situation (see above) and apparently a fix is planned for a future FW release. we'll see how that one pans out.

unfortunately, that will take a while and won't address drives already deployed [with legacy FW]. therefore i think i'd like to take you up on the idea of adding an option to cap `MaxComPacketSize`. we're only interested in the library side of things, not the binaries. at the very least we'll need to add an option to `core.NewControlSession()` (`ControlSessionOpt`), something like `WithMaxComPacketSize(uint)`, but i also suspect an `InitializeOpt` will be required for `locking.Initialize()`. AFAICT this should be backwards compatible with the existing library users, and should allow us to address the immediate problem - albeit in a somewhat heavy-handed fashion.

would such a change be acceptable for you? if so - we can whip up a corresponding separate PR.

ChriMarMe replied:
It sounds reasonable to introduce that. There is a real world problem and you need it fixed. Also your proposed solution follows a pattern already used and it is optional for any user.
